### PR TITLE
Fix buffer overflow handling for string_body and mutable_body

### PR DESCRIPTION
**Pull request by octopus-prime**
issue #553

### CODE REVIEW

**Comment by vinniefalco** [comment]
It needs to be all squashed down into one commit

File: include/beast/http/string_body.hpp [hunk 1]
```diff
@@ -117,14 +120,14 @@ struct string_body
...
             }
-            catch(std::length_error const&)
+            catch(std::exception const&)
             {
                 ec = error::buffer_overflow;
                 return;
             }
-            ec.assign(0, ec.category());
             buffer_copy(boost::asio::buffer(
                 &body_[0] + len, n), buffers);
```
Comment (vinniefalco) [inline include/beast/http/string_body.hpp]: if `std::distance(buffers.begin(), buffers.end()==1` then using `body_.append` is more efficient

**Comment by octopus-prime** [comment]
> It needs to be all squashed down into one commit

Ouch.
Closing this pull request?
New branch from master again.
All changes.
One commit.
New pull request?

File: include/beast/http/string_body.hpp (hunk 1 above)
Comment (octopus-prime) [inline include/beast/http/string_body.hpp]: I would always call append.
The container should know how to handle allocation and copying at best...
Better than we know...
So resize() would be obsolete too. But people told me, that resize() is VERY important.

Or do you mean an optimization for 1 character only?! Like push_back() ?

**Comment by vinniefalco** [comment]
What? No! You can squash it from the command line. Try
```
git remote add upstream git@github.com:vinniefalco/Beast.git
git rebase -i upstream/master
```

or you can use 
```
git reset 1e8be328
git add -A
git commit --amend
```

When you push to your remote use
```
git push -f
```

**Comment by octopus-prime** [comment]
Do i need
`git add -A`
?

There are no new files to commit.
But i have new files in workspace, e.g. `.project`

**Comment by vinniefalco** [comment]
I don't know what `.project` is. If you used `git reset` you should have modified files in you working directory. Try resetting back to your original branch and then do the interactive rebase  (`git rebase -i`)

**Comment by vinniefalco** [comment]
See https://github.com/vinniefalco/Beast/commits/op

**Comment by octopus-prime** [comment]
WTH?! git push did close the pull request?! I get crazy!!

**Comment by octopus-prime** [comment]
okay. shit happens. i will try again tomorrow...

**Comment by vinniefalco** [comment]
No worries, take your time and have patience, this will work.

**Pull request closed without merging**